### PR TITLE
Add call semantics checks, pass call04.f90 test

### PR DESCRIPTION
**Pull request by klausler**
Here's some more call semantics checks.  Note that this branch is a branch of "pmk-calls", and that I have not yet merged that branch with master.

### CODE REVIEW

File: lib/evaluate/characteristics.cc [hunk 1]
```diff
@@ -310,6 +331,51 @@ std::optional<DummyArgument> DummyArgument::Characterize(
...
+std::optional<DummyArgument> DummyArgument::FromActual(
+    std::string &&name, const Expr<SomeType> &expr, FoldingContext &context) {
+  return std::visit(
+      common::visitors{
+          [&](const BOZLiteralConstant &) {
+            return std::make_optional<DummyArgument>(std::move(name),
+                DummyDataObject{
+                    TypeAndShape{DynamicType::TypelessIntrinsicArgument()}});
+          },
+          [&](const NullPointer &) { return std::optional<DummyArgument>{}; },
```
Comment (psteinfeld) [inline lib/evaluate/characteristics.cc]: I think that the "&" reference capture is not needed here.

File: lib/evaluate/shape.cc [hunk 2]
```diff
@@ -349,8 +353,28 @@ Shape GetUpperBounds(FoldingContext &context, const NamedEntity &base) {
...
+              Shape shape;
+              int n{object.shape().Rank()};
+              NamedEntity base{symbol};
+              for (int dimension{0}; dimension < n; ++dimension) {
+                shape.emplace_back(GetExtent(context_, base, dimension));
+              }
+              return Result{shape};
+            }
+          },
+          [&](const semantics::EntityDetails &) {
```
Comment (psteinfeld) [inline lib/evaluate/shape.cc]: The reference capture specification is not needed here or below for the default case.

File: lib/evaluate/shape.cc [hunk 3]
```diff
@@ -451,21 +468,80 @@ auto GetShapeHelper::operator()(const ProcedureRef &call) const -> Result {
...
+        if (auto shape{(*this)(call.arguments().at(0))}) {
+          if (shape->size() == 2) {
+            std::swap((*shape)[0], (*shape)[1]);
+            return shape;
+          }
+        }
       }
+    } else if (intrinsic->characteristics.value().attrs.test(characteristics::
+                       Procedure::Attr::NullPointer)) {  // NULL(MOLD=)
+      return (*this)(call.arguments());
```
Comment (psteinfeld) [inline lib/evaluate/shape.cc]: Should there be curly braces around this "return"?

File: lib/semantics/check-declarations.cc [hunk 4]
```diff
@@ -88,31 +88,44 @@ void CheckHelper::Check(Symbol &symbol) {
...
+                return IsCoarray(symbol) && IsAllocatable(symbol);
+              }})) {  // C846
+        messages_.Say(
+            "An INTENT(OUT) dummy argument may not be, or contain, an ALLOCATABLE coarray"_err_en_US);
+      }
+      if (IsOrContainsEventOrLockComponent(symbol)) {  // C847
+        messages_.Say(
+            "An INTENT(OUT) dummy argument may not be, or contain, EVENT_TYPE or LOCK_TYPE"_err_en_US);
+      }
+    }
```
Comment (psteinfeld) [inline lib/semantics/check-declarations.cc]: Thanks for adding these checks!  They were on my list.

File: lib/evaluate/shape.cc (hunk 3 above)
Comment (klausler) [reply lib/evaluate/shape.cc]: They are present.

File: lib/semantics/check-declarations.cc [hunk 5]
```diff
@@ -88,31 +88,44 @@ void CheckHelper::Check(Symbol &symbol) {
...
     }
   }
   if (auto *object{symbol.detailsIf<ObjectEntityDetails>()}) {
     Check(object->shape());
     Check(object->coshape());
+    if (object->isDummy() && symbol.attrs().test(Attr::INTENT_OUT)) {
+      if (FindUltimateComponent(symbol,
+              std::function<bool(const Symbol &)>{[](const Symbol &symbol) {
+                return IsCoarray(symbol) && IsAllocatable(symbol);
+              }})) {  // C846
```
Comment (tskeith) [inline lib/semantics/check-declarations.cc]: Why not just: `if (FindUltimateComponent(symbol, [](const Symbol &symbol) {...}))`?

Comment (klausler) [reply lib/semantics/check-declarations.cc]: Sure.

**Pull request merged**